File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/InvestigationActivity.java
package com.TritiumGaming.phasmophobiaevidencepicker.activities;

import android.os.Bundle;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;
import androidx.navigation.ui.NavigationUI;

import com.TritiumGaming.phasmophobiaevidencepicker.R;
import com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.EvidenceViewModel;
import com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.GlobalPreferencesViewModel;
import com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.MapMenuViewModel;
import com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.ObjectivesViewModel;
import com.google.android.material.navigation.NavigationBarView;
import com.google.android.material.navigation.NavigationView;

/**
 * InvestigationActivity class
 *
 * @author TritiumGamingStudios
 */
public class InvestigationActivity extends PETActivity {

    protected GlobalPreferencesViewModel globalPreferencesViewModel;

    protected EvidenceViewModel evidenceViewModel;
    protected ObjectivesViewModel objectivesViewModel;
    protected MapMenuViewModel mapMenuViewModel;

    public DrawerLayout drawerLayout;
    public ActionBarDrawerToggle actionBarDrawerToggle;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_investigation_solo);

    }

    public void initComponents() {

        NavHostFragment navHostFragment =
                (NavHostFragment) getSupportFragmentManager().findFragmentById(R.id.fragment_container);

        Log.d("NavHostFragment", navHostFragment == null ? "null" : "not null");
        if(navHostFragment == null) { return; }

        NavController navController = navHostFragment.getNavController();

        NavigationBarView navBarView = findViewById(R.id.item_navigation_bar);
        Log.d("Bar", navBarView == null ? "null" : "not null");
        setNavigationBarBehavior(navBarView, navController);

        NavigationView navDrawerView = findViewById(R.id.layout_navigation_drawer_view);
        Log.d("Drawer", navDrawerView == null ? "null" : "not null");
        setNavigationDrawer(navDrawerView, navController);

    }

    protected ViewModelProvider.AndroidViewModelFactory initViewModels() {

        ViewModelProvider.AndroidViewModelFactory factory = super.initViewModels();

        evidenceViewModel = factory.create(
                EvidenceViewModel.class);

        objectivesViewModel = factory.create(
                ObjectivesViewModel.class);

        mapMenuViewModel = factory.create(
                MapMenuViewModel.class);

        return factory;
    }

    protected void initPrefs() {
        super.initPrefs();

        setLanguage(getAppLanguage());
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions,
                                           int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        recreate();
    }

    public void clearViewModels() {
        if(objectivesViewModel != null) {
            objectivesViewModel.reset();
        }
        if(evidenceViewModel != null) {
            evidenceViewModel.reset();
        }
    }


    private void setNavigationDrawer(NavigationView navView, NavController navController) {
        drawerLayout = findViewById(R.id.layout_navigation_drawer);

        if(drawerLayout == null) { return; }

        actionBarDrawerToggle = new ActionBarDrawerToggle(this, drawerLayout, R.string.navigation_open_state, R.string.navigation_closed_state);

        // pass the Open and Close toggle for the drawer layout listener
        // to toggle the button
        drawerLayout.setDrawerLockMode(DrawerLayout.LOCK_MODE_LOCKED_CLOSED);
        drawerLayout.addDrawerListener(actionBarDrawerToggle);
        actionBarDrawerToggle.syncState();

        // to make the Navigation drawer icon always appear on the action bar
        //getActivity().getSupportFragmentManager().setDisplayHomeAsUpEnabled(true);
        drawerLayout.closeDrawer(GravityCompat.START);

        setNavigationDrawerBehavior(navView, navController);
    }

    private void setNavigationDrawerBehavior(NavigationView navView, NavController navController) {

        if(drawerLayout != null) {

            if(navView != null) {
                NavigationUI.setupWithNavController(navView, navController);

                navView.setNavigationItemSelectedListener(item -> {

                    boolean success = NavigationUI.onNavDestinationSelected(item, navController);
                    if(success) {
                        drawerLayout.closeDrawer(GravityCompat.START);

                        return true;
                    }

                    if(drawerLayout != null) {
                        if (item.getItemId() == (R.id.action_home)) {
                            finish();
                        }
                    }

                    return true;
                });
            }
        }
    }

    public void setNavigationBarBehavior(NavigationBarView navView, NavController navController) {

        if(navView != null) {
            NavigationUI.setupWithNavController(navView, navController);
            navView.setOnItemSelectedListener(item -> {

                Log.d("Bar", "item: " + item.getTitle());

                if(NavigationUI.onNavDestinationSelected(item, navController)) {
                    if(drawerLayout != null) {

                        drawerLayout.closeDrawer(GravityCompat.START);
                    }

                    return true;
                }

                if (item.getItemId() == (R.id.open_menu)) {
                    if(drawerLayout != null) {
                        if (drawerLayout.isOpen()) {
                            drawerLayout.closeDrawer(GravityCompat.START);
                        } else {
                            drawerLayout.openDrawer(GravityCompat.START);
                        }
                    }
                }
                return true;

            });
        }
    }


    public boolean closeDrawer() {
        if(drawerLayout != null) {
            if (drawerLayout.isOpen()) {
                drawerLayout.closeDrawer(GravityCompat.START);
                return true;
            }
        }
        return false;
    }

    @Override
    protected void onStart() {

        super.onStart();
    }

    /**
     * Resets ObjectiveViewModel and EvidenceViewModel data upon Activity exit
     */
    @Override
    public void onBackPressed() {

        clearViewModels();

        super.onBackPressed();
    }

    @Override
    public void finish() {

        clearViewModels();

        super.finish();
    }

}

File: app-legacy/src/main/java/com/tritiumgaming/phasmophobiaevidencepicker/presentation/ui/activities/investigation/InvestigationActivity.kt
package com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.investigation

import android.os.Bundle
import android.util.Log
import android.view.MenuItem
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.lifecycle.ViewModelProvider.AndroidViewModelFactory
import androidx.navigation.NavController
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.ui.NavigationUI.onNavDestinationSelected
import androidx.navigation.ui.NavigationUI.setupWithNavController
import com.google.android.material.navigation.NavigationBarView
import com.google.android.material.navigation.NavigationView
import com.tritiumgaming.phasmophobiaevidencepicker.R
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.pet.PETActivity
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.viewmodel.InvestigationViewModel
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.viewmodel.MapMenuViewModel
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.viewmodel.ObjectivesViewModel

/**
 * InvestigationActivity class
 *
 * @author TritiumGamingStudios
 */
open class InvestigationActivity : PETActivity() {
    protected var investigationViewModel: InvestigationViewModel? = null
    protected var objectivesViewModel: ObjectivesViewModel? = null
    protected var mapMenuViewModel: MapMenuViewModel? = null

    private var drawerLayout: DrawerLayout? = null
    private var actionBarDrawerToggle: ActionBarDrawerToggle? = null

    private var navigationBarView: NavigationBarView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_investigation)
    }

    override fun initViewModels(): AndroidViewModelFactory? {
        val factory: AndroidViewModelFactory? = super.initViewModels()
        factory?.let {
            initInvestigationViewModel(factory)
            initObjectivesViewModel(factory)
            initMapMenuViewModel(factory)
        }

        return factory
    }

    private fun initMapMenuViewModel(factory: AndroidViewModelFactory) {
        mapMenuViewModel = factory.create(MapMenuViewModel::class.java)
    }

    private fun initObjectivesViewModel(factory: AndroidViewModelFactory) {
        objectivesViewModel = factory.create(ObjectivesViewModel::class.java)
    }

    private fun initInvestigationViewModel(factory: AndroidViewModelFactory) {
        investigationViewModel = factory.create(InvestigationViewModel::class.java)
    }

    override fun loadPreferences() {
        super.loadPreferences()
        globalPreferencesViewModel?.currentLanguageAbbr?.let { currentLangAbbr ->
            setLanguage(currentLangAbbr)
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        recreate()
    }

    private fun resetViewModels() {
        objectivesViewModel?.reset()
        investigationViewModel?.reset()
    }

    fun initNavigationComponents() {
        val navHostFragment =
            supportFragmentManager.findFragmentById(R.id.fragment_container) as NavHostFragment?

        Log.d("NavHostFragment", if (navHostFragment == null) "null" else "not null")

        navHostFragment?.navController?.let { navController ->
            val navBarView = findViewById<NavigationBarView>(R.id.item_navigation_bar)
            Log.d("Bar", if (navBarView == null) "null" else "not null")
            setNavigationBarBehavior(navBarView, navController)

            val navDrawerView = findViewById<NavigationView>(R.id.layout_navigation_drawer_view)
            Log.d("Drawer", if (navDrawerView == null) "null" else "not null")
            buildNavigationDrawer(navDrawerView, navController)
        }

    }

    private fun buildNavigationDrawer(navView: NavigationView?, navController: NavController) {
        drawerLayout = findViewById(R.id.layout_navigation_drawer)

        drawerLayout?.let { drawerLayout ->
            actionBarDrawerToggle = ActionBarDrawerToggle(
                this, drawerLayout,
                R.string.navigation_open_state,
                R.string.navigation_closed_state
            )

            // pass the Open and Close toggle for the drawer layout listener
            // to toggle the button
            actionBarDrawerToggle?.let { actionBarDrawerToggle ->
                drawerLayout.setDrawerLockMode(DrawerLayout.LOCK_MODE_LOCKED_CLOSED)
                drawerLayout.addDrawerListener(actionBarDrawerToggle)
                actionBarDrawerToggle.syncState()
            }

            // to make the Navigation drawer icon always appear on the action bar
            //getActivity().getSupportFragmentManager().setDisplayHomeAsUpEnabled(true);
            drawerLayout.closeDrawer(GravityCompat.START)
        }

        setNavigationDrawerBehavior(navView, navController)
    }

    private fun setNavigationDrawerBehavior(
        navView: NavigationView?, navController: NavController) {
        navView?.let {
            setupWithNavController(navView, navController)
            navView.setNavigationItemSelectedListener(NavigationView.OnNavigationItemSelectedListener { item: MenuItem ->
                drawerLayout?.let { drawerLayout ->
                    if (onNavDestinationSelected(item, navController)) {
                        drawerLayout.closeDrawer(GravityCompat.START)
                        return@OnNavigationItemSelectedListener true
                    }
                    if (item.itemId == (R.id.action_home)) {
                        finish()
                    }
                }
                true
            })
        }
    }

    private fun setNavigationBarBehavior(navView: NavigationBarView?, navController: NavController) {
        navView?.let { nv ->
            navigationBarView = nv

            setupWithNavController(nv, navController)

            nv.setOnItemSelectedListener(
                NavigationBarView.OnItemSelectedListener { item: MenuItem ->
                if (onNavDestinationSelected(item, navController)) {
                    drawerLayout?.closeDrawer(GravityCompat.START)

                    return@OnItemSelectedListener true
                }
                if (item.itemId == (R.id.open_menu)) {
                    drawerLayout?.let { drawerLayout ->
                        when (drawerLayout.isOpen) {
                            true -> drawerLayout.closeDrawer(GravityCompat.START)
                            false -> drawerLayout.openDrawer(GravityCompat.START)
                        }
                    }
                }
                true
            })

            nv.itemIconTintList = null
        }
    }


    fun closeNavigationDrawer(): Boolean {
        drawerLayout?.let { drawerLayout ->
            if (drawerLayout.isOpen) {
                drawerLayout.closeDrawer(GravityCompat.START)
                return true
            }
        }
        return false
    }

    override fun onBackPressed() {
        resetViewModels()

        super.onBackPressed()
    }

    override fun finish() {
        resetViewModels()

        super.finish()
    }
}
